test_pull_requests

```diff
@@ -152,6 +185,115 @@ def submitted_tests_for_branch(branch)
   $submitted_tests[branch] ? $submitted_tests[branch] : $submitted_tests['*']
 end
 
+# Helper function to create appropriate regex for matching/extracting
+# :build_url field from bot comment
+def build_url_field_regex
+  if Properties &&
+    / \((#{Regexp::escape(Properties['jenkins_url'])}[^)]+)\)/
+  else
+    # Use this test to set the build url matching pattern
+    if ENV['BUILD_URL']
+      # Don't set build_url directly from the env var, since we want
+      # to make sure they match
+      / \((#{Regexp::escape(ENV['BUILD_URL'])})\)/
+    else
+      / \((https?:[^)]+)\)/
+    end
+  end
+end
+
+# Helper function to pull the test prefix from the settings object or
+# try to determine it from the provided bot_comment_body
+def extract_test_prefix(bot_comment_body, settings=nil)
+  if settings
+    settings['test_prefix']
+  else
+    # Do our best to find the prefix since we don't know it beforehand
+    if bot_comment_body =~ /^([^:(]+): /
+      $1.strip
+    elsif bot_comment_body =~ /^([^ (]+) /
+      $1.strip
+    end
+  end
+end
+
+# Pulls out as much metadata from the bot comment as it can, including
+# state (AKA content), Jenkins build URL, base repo commit ID, and the
+# extended tests metadata
+def extract_bot_comment_fields(bot_comment_body, settings=nil)
```

Comment:
@stevekuznetsov 

> The fields remain optional -- if you have the values and want us to take advantage of the fact that you have the values, pass them in.

I meant more work in terms of the interpreter. It has to pass 2 references instead of 1

> Sure, but not the `settings` that also contains things like queue information, team definitions, flake identification...

But we're only passing a reference, we're not doing a deep copy

> 
> ...`settings` could be non-nil but `settings['test_prefix']` could be nil.
> 

That shouldn't happen, and if we're really worried about that we should add config validation instead of testing for required values whenever we use them.

> Passing around a massive data structure with completely unrelated information for convenience smells bad.

Unpacking values that are tightly-coupled and validating them both at run time when we could just check if their containing data structure is `nil` smells bad.